packages/fxa-settings/src/pages/PostVerify/ThirdPartyAuthCallback/index.tsx

```diff
@@ -3,40 +3,150 @@
+      const redirect = stateParams.get('redirect_uri');
+      // if the state contains a redirect_uri, we need to redirect to RP
+      // otherwise we redirect internally
+      if (redirect) {
+        const url = new URL(redirect);
+        // TODO append other params from state to the redirect URL
+        return url;
+      }
+    }
+    return undefined;
+  };
+
+  // auth params are received from the third party auth provider
+  // and are required to verify the account
+  const getAuthParams = () => {
+    const code = params.get('code');
+    const providerFromParams = params.get('provider');
+    let provider: AUTH_PROVIDER | undefined;
+    if (providerFromParams === 'apple') {
+      provider = AUTH_PROVIDER.APPLE;
+    } else {
+      provider = AUTH_PROVIDER.GOOGLE;
+    }
+
+    return { code, provider };
+  };
 
-        navigate(redirectUrl, { replace: true });
+  const handlePostSignInNavigation = () => {
+    // TODO ensure correct redirects for all integrations (OAuth, Desktop, Mobile)
+    // redirect is constructed from state param in the URL params
+    const redirectURL = getRedirectUrl();
+    if (redirectURL) {
+      // get the stashed state with origin information
+      // use it to reconstruct redirect for oauth
+      // the state is the entire URL of the origin and includes the redirect_uri in a param
+      // if authenticating from a RP
+      const state = params.get('state');
+      if (state) {
+        // we may need to deconstruct the state to access/modify the redirect URL
+        const stateParams = new URL(decodeURIComponent(state)).searchParams;
+        const redirect = stateParams.get('redirect_uri');
+        // if the state contains a redirect_uri, we need to redirect to RP
+        // otherwise we redirect internally
+        if (redirect) {
+          hardNavigate(redirect);
+        } else {
+          // general redirect to settings for non-RP
+          // currently, redirect to /settings fails with an "unauthenticated" error from GQL
+          // and redirects to /signin where signin in *again* with ThirdPArty Auth successfully
+          // navigates to /settings
+          // however, after signup we can see the account is successfully created in the db
+          // and added to local storage
+          // need to determine what is missing to directly be considered signed in
+          // when signup with thrid party auth is handled with react from end to end
+          navigate(`/settings${window.location.search}`);
+        }
       }
-    } catch (e) {
-      // noop. navigate to home below.
     }
+  };
 
-    navigate('/', { replace: true });
+  // Persist account data to local storage to match parity with content-server
+  // this allows the recent account to be used for /signin
+  const storeAccountData = (accountData: StoredAccountData) => {
+    persistAccount(accountData);
+    setCurrentAccount(accountData.uid);
+    sessionToken(accountData.sessionToken);
```

Comment:
Would we also need to update graphql?

```
    await this.apolloClient.cache.writeQuery({
      query: GET_LOCAL_SIGNED_IN_STATUS,
      data: { isSignedIn: true },
    });
```

Reply:
I had added this in `verifyAccountThirdParty`, but I think it needs to be reshuffled. We're missing the actual sign in that will be completed in FXA-8834/FXA-6488 and that's where it should go. Added a note.